File: apps/mobile/lib/core/logic/binder_move.dart
import '../models/binder.dart';
import '../models/binder_entry.dart';
import '../models/card_model.dart';

/// Result of [moveBinderCopies]. [entries] is the post-move list (unchanged on
/// refusal). [distinctOwned] is unique owned printing ids.
class BinderMoveResult {
  const BinderMoveResult({
    required this.ok,
    this.reason,
    required this.entries,
    required this.distinctOwned,
  });

  final bool ok;
  final String? reason;
  final List<BinderEntry> entries;
  final int distinctOwned;
}

int distinctOwnedCount(Iterable<BinderEntry> entries) => entries
    .where((e) => !e.isWanted && e.quantity > 0)
    .map((e) => e.card.id)
    .toSet()
    .length;

/// Move [quantity] copies of a Printing+condition from one live Binder to
/// another. Never refuses for the shared `binderCards` cap. Want List is never
/// a destination.
BinderMoveResult moveBinderCopies({
  required List<Binder> binders,
  required List<BinderEntry> entries,
  required String fromBinderId,
  required String toBinderId,
  required String printingId,
  required String condition,
  required int quantity,
  DateTime? now,
}) {
  final stamp = now ?? DateTime.now();
  int owned() => distinctOwnedCount(entries);

  BinderMoveResult refuse(String reason) => BinderMoveResult(
        ok: false,
        reason: reason,
        entries: entries,
        distinctOwned: owned(),
      );

  if (quantity < 1) return refuse('invalid-quantity');
  if (fromBinderId == toBinderId) return refuse('same-binder');
  if (toBinderId == 'want' || toBinderId.trim().isEmpty) {
    return refuse('invalid-destination');
  }

  Binder? dest;
  for (final b in binders) {
    if (b.clientId == toBinderId) {
      dest = b;
      break;
    }
  }
  if (dest == null || !dest.isLive) return refuse('invalid-destination');

  final srcIdx = entries.indexWhere((e) =>
      !e.isWanted &&
      e.card.id == printingId &&
      e.resolvedBinderId == fromBinderId &&
      e.condition == condition);
  if (srcIdx < 0) return refuse('missing-source');
  final source = entries[srcIdx];
  if (quantity > source.quantity) return refuse('invalid-quantity');

  final next = [...entries];
  final remaining = source.quantity - quantity;
  if (remaining <= 0) {
    next.removeAt(srcIdx);
  } else {
    next[srcIdx] = source.copyWith(quantity: remaining);
  }

  final destIdx = next.indexWhere((e) =>
      !e.isWanted &&
      e.card.id == printingId &&
      e.resolvedBinderId == toBinderId &&
      e.condition == condition);
  if (destIdx >= 0) {
    next[destIdx] =
        next[destIdx].copyWith(quantity: next[destIdx].quantity + quantity);
  } else {
    next.add(BinderEntry(
      card: source.card,
      quantity: quantity,
      condition: condition,
      isWanted: false,
      binderId: toBinderId,
      addedAt: stamp,
    ));
  }

  return BinderMoveResult(
    ok: true,
    entries: next,
    distinctOwned: distinctOwnedCount(next),
  );
}

/// Contract-shaped row used only by tests. Production passes [BinderEntry].
BinderEntry binderEntryFromContract(Map<String, dynamic> row, {DateTime? now}) {
  final wanted = row['isWanted'] as bool? ?? false;
  return BinderEntry(
    card: CardModel(
      id: row['printingId'] as String,
      name: (row['name'] as String?) ?? row['printingId'] as String,
      tcgMarket: (row['tcgMarket'] as num?)?.toDouble(),
    ),
    quantity: (row['quantity'] as num?)?.toInt() ?? 1,
    condition: (row['condition'] as String?) ?? 'NM',
    isWanted: wanted,
    binderId: wanted ? null : row['binderId'] as String?,
    addedAt: now ?? DateTime.now(),
  );
}

Binder binderFromContract(Map<String, dynamic> row, {DateTime? now}) {
  final stamp = now ?? DateTime.now();
  final deleted = row['deletedAt'] as String?;
  return Binder(
    clientId: row['clientId'] as String,
    name: (row['name'] as String?) ?? row['clientId'] as String,
    role: BinderRole.parse(row['role'] as String?),
    createdAt: stamp,
    updatedAt: stamp,
    deletedAt: deleted == null ? null : DateTime.tryParse(deleted),
  );
}

File: apps/web/src/utils/binderMove.js
import { distinctOwnedCount } from './freeLimits.js';

/**
 * Move copies of a Printing+condition between live Binders.
 * Golden cases: packages/contracts/binder_move.json
 * Never refuses for binderCards. Want List is never a destination.
 *
 * @returns {{ ok: boolean, reason: string|null, entries: Array, distinctOwned: number }}
 */
export function moveBinderCopies({
    binders,
    entries,
    fromBinderId,
    toBinderId,
    printingId,
    condition = 'NM',
    quantity,
}) {
    const snapshot = (entries || []).map((e) => ({ ...e }));
    const owned = () => distinctOwnedCount(snapshot);
    const refuse = (reason) => ({
        ok: false,
        reason,
        entries: snapshot,
        distinctOwned: owned(),
    });

    const qty = Number(quantity);
    if (!Number.isFinite(qty) || qty < 1) return refuse('invalid-quantity');
    if (fromBinderId === toBinderId) return refuse('same-binder');
    if (!toBinderId || toBinderId === 'want') return refuse('invalid-destination');

    const dest = (binders || []).find((b) => b.clientId === toBinderId);
    if (!dest || dest.deletedAt) return refuse('invalid-destination');

    const srcIdx = snapshot.findIndex((e) =>
        !e.isWanted &&
        (e.printingId || e.cardId) === printingId &&
        (e.binderId || null) === fromBinderId &&
        (e.condition || 'NM') === condition,
    );
    if (srcIdx < 0) return refuse('missing-source');
    const source = snapshot[srcIdx];
    if (qty > (Number(source.quantity) || 0)) return refuse('invalid-quantity');

    const remaining = (Number(source.quantity) || 0) - qty;
    if (remaining <= 0) snapshot.splice(srcIdx, 1);
    else snapshot[srcIdx] = { ...source, quantity: remaining };

    const destIdx = snapshot.findIndex((e) =>
        !e.isWanted &&
        (e.printingId || e.cardId) === printingId &&
        (e.binderId || null) === toBinderId &&
        (e.condition || 'NM') === condition,
    );
    if (destIdx >= 0) {
        snapshot[destIdx] = {
            ...snapshot[destIdx],
            quantity: (Number(snapshot[destIdx].quantity) || 0) + qty,
        };
    } else {
        snapshot.push({
            ...source,
            binderId: toBinderId,
            isWanted: false,
            quantity: qty,
            condition,
            printingId: source.printingId || source.cardId || printingId,
            cardId: source.cardId || source.printingId || printingId,
        });
    }

    return {
        ok: true,
        reason: null,
        entries: snapshot,
        distinctOwned: distinctOwnedCount(snapshot),
    };
}
